### PR TITLE
fix cached_path for hub downloads

### PR DESCRIPTION
**Pull request by epwalsh**
Fixes issue described [here](https://github.com/allenai/allennlp/pull/5109#issuecomment-824335671), and also ensures the downloads end up in the right cache directory, with meta ".json" files that adhere to our spec.

### CODE REVIEW

**Comment by epwalsh** [comment]
@dirkgr and @LysandreJik, please take a look when you get a chance

File: allennlp/common/file_utils.py [hunk 1]
```diff
@@ -476,7 +457,7 @@ def _http_etag(url: str) -> Optional[str]:
     with _session_with_backoff() as session:
         response = session.head(url, allow_redirects=True)
     if response.status_code != 200:
-        raise IOError(
+        raise OSError(
```
Comment (epwalsh) [inline allennlp/common/file_utils.py:460]: This was a separate bug not related to HF Hub stuff.

**Review by dirkgr — APPROVED** [review]
LGTM, but I'd love it if @LysandreJik could take a look.

File: allennlp/common/file_utils.py [hunk 2]
```diff
@@ -208,15 +206,34 @@ def cached_path(
...
-    determine which. If it's a URL, download the file and cache it, and
-    return the path to the cached file. If it's already a local path,
-    make sure the file exists and then return the path.
+    Given something that might be a URL or local path, determine which.
+    If it's a remote resource, download the file and cache it, and
+    then return the path to the cached file. If it's already a local path,
+    make sure the file exists and return the path.
+
+    For URLs, "http://", "https://", "s3://", and "hf://" are all supported.
+    The later corresponds to the HuggingFace Hub.
```
Comment (LysandreJik) [inline allennlp/common/file_utils.py:215]:
```suggestion
    The latter corresponds to the HuggingFace Hub.
```

File: allennlp/common/file_utils.py [hunk 3]
```diff
@@ -867,13 +895,37 @@ def get_from_cache(url: str, cache_dir: Union[str, Path] = None) -> str:
...
+        if identifier.count("/") > 1:
+            filename = "/".join(identifier.split("/")[2:])
+            model_identifier = "/".join(identifier.split("/")[:2])
+            return _hf_hub_download(url, model_identifier, filename, cache_dir)
+        elif identifier.count("/") == 1:
+            # 'hf://' URLs like 'hf://xxxx/yyyy' are potentially ambiguous,
+            # because this could refer to either:
+            #  1. the file 'yyyy' in the 'xxxx' repository, or
+            #  2. the repo 'yyyy' under the user/org name 'xxxx'.
+            # We default to (1), but if we get a 404 error then we try (2).
```
Comment (LysandreJik) [inline allennlp/common/file_utils.py:911]: Very nice catch, I'll have to keep that in mind.

**Review by LysandreJik — APPROVED** [review]
This is great, thank you for tying it into the rest of your code.

**Pull request merged**